File: Python/HeapSort.py
#Heap Sort,


def Heapy(DataStore,size,i):
    largest = i
    left = 2*i+1
    right = 2*i+2

    if(left<size and DataStore[left]>=DataStore[largest]):
        largest = left
    
    if(right<size and DataStore[right]>=DataStore[largest]):
        largest = right

    if(largest !=i):
        temp  = DataStore[i]
        DataStore[i] = DataStore[largest]
        DataStore[largest] = temp

        Heapy(DataStore,size,largest)


#Heap Sort Method, 
def HeapSort(DataStore,size):
    #max Heap, 
    
    for x in range(size//2,-1,-1):
        Heapy(DataStore,size,x)

    #sort, 
    for y in range(size-1,0,-1):
        temp  = DataStore[0]
        DataStore[0] = DataStore[y]
        DataStore[y] = temp

        Heapy(DataStore,y,0)



DataStore = [1,12,9,5,6,10]
size = len(DataStore)
print("UnSort DataStore : ",DataStore)
HeapSort(DataStore,size)
print("  Sort DataStore : ",DataStore)

File: Java/HeapSort.java
//Heap Sort,

public class HeapSort{


    //create heapy, 
    public static void heapy(int[] DataStore,int size,int i){
        int largest = i;
        int left = 2*i+1;
        int right = 2*i+2;

        if(left<size && DataStore[left]>=DataStore[largest]){
            largest = left; 
        }

        if(right<size && DataStore[right]>=DataStore[largest]){
            largest = right;
        }

        if(largest != i){
            int temp = DataStore[i];
            DataStore[i] = DataStore[largest];
            DataStore[largest] = temp;

            heapy(DataStore, size, largest);
        }
    }



    //Heap Sort Method, 
    public static void heapsort(int[] DataStore, int size){
        //max heap,
        for(int i=size/2-1; i>=0; i--){
            heapy(DataStore, size, i);
        }

        // Sort, 
        for(int i=size-1; i>=0; i--){
            int temp = DataStore[0];
            DataStore[0] = DataStore[i];
            DataStore[i] = temp;

            heapy(DataStore,i,0);
        }
    }


    //print DataStore,
    public static void ShowData(int[] DataStore,int size){
        for(int a=0; a<size; a++){
            System.out.print(DataStore[a] + " ");
        }
        System.out.println(" ");
    }

    public static void main(String[] args) {
        int[] DataStore = {1,12,9,5,6,10};
        int size = DataStore.length;
        System.out.print("UnSort DataStore : ");
        ShowData(DataStore, size);
        heapsort(DataStore, size);
        System.out.print("  Sort DataStore : ");
         ShowData(DataStore, size);
    }
}